File: CPP/Algorithm/Meet_in_the_Middle_(Knapsack).cpp
// Complexity: O(2^(n/2) * log(2^(n/2)))
// const int64_t N = 40 + 5;

int64_t aa[N];
auto ssum(auto s, auto e) {
    int64_t n = (1ll << (e - s)); vector<int64_t> res(n);
    for (int i = 0; i < n; ++i) {
        int64_t sum = 0, hi = __builtin_popcount(i);
        for (j = 0; hi && j < e - s; ++j) {
            if (i & (1ll << j)) { sum += aa[j + s]; --hi; }
        }
        res[i] = sum;
    }
    return res;
}

// Operation:
    int64_t sum; cin >> n >> sum;
    for (i = 0; i < n; ++i) { cin >> aa[i]; }
    auto h1 = ssum(0, n >> 1), h2 = ssum(n >> 1, n);
    sort(begin(h2), end(h2));
    int64_t c = 0, sz = h1.size();
    for (i = 0; i < sz; ++i) {
        if (sum >= h1[i])
            c += upper_bound(begin(h2), end(h2), sum - h1[i]) - lower_bound(begin(h2), end(h2), sum - h1[i]);
    }
    cout << c;

// Problems:
// https://cses.fi/problemset/task/1628/
// https://atcoder.jp/contests/abc184/tasks/abc184_f

File: Python/Algorithm/Meet_in_the_Middle_(Knapsack).py
# Complexity: O(2^(n/2) * log(2^(n/2)))

from bisect import bisect_left
from bisect import bisect_right
def ssum(ls, s, e):
    n = 1 << (e - s); res = []
    for i in range(n):
        sum = 0
        for j in range(e - s):
            if i & (1<<j): sum += ls[j+s]
        res.append(sum)
    return res

def cnt(ls, tot):
    sz = len(ls)
    l = ssum(ls, 0, sz >> 1); r = ssum(ls, sz >> 1, sz)
    r.sort(); c, n = 0, len(l)
    for i in range(n):
        if tot >= l[i]:
            lo = bisect_left(r, tot-l[i]); hi = bisect_right(r, tot-l[i])
            c += hi - lo
    return c

# Problems:
# https://cses.fi/problemset/task/1628/
# https://atcoder.jp/contests/abc184/tasks/abc184_f